snarePip
===============================
[![Build Status](https://travis-ci.com/huqiwen0313/snarePip.svg?branch=main)](https://travis-ci.com/github/huqiwen0313/snarePip)

## Overview
snarePip is an analysis pipeline designed for [snare-seq]( https://www.nature.com/articles/s41587-019-0290-0) data. 
It can also be used for joint processing and analyze single-cell RNA and atac sequencing dataset. 

It contains 2 independent modules: data processing/analysis and meta-table manipulation.
The data processing part is based on [Snakemake](https://snakemake.readthedocs.io) framework, which provides an automated framework for complex analysis including quality assessment, doublet removal, cell clustering and identification, peak generation, differential accessible region identification and linkage analysis. 

The meta-table part is designed for easier meta-table manipulation and migration including automated extraction of sample information, preparation of data uploading and update of QC statics generated by the pipeline. 
The current version of pipeline automatically processing single-cell RNA and ATAC datasets with flexible analysis modules and generate summary reports for both quality assessment and downstream analysis.

The entire framework is based on Directed acyclic graph (DAG) and [luigi](https://luigi.readthedocs.io/en/stable/) framework is used to build the DAG that connect data processing and meta-table part together with salted workflow for better version control.

<p align="center"> 
<img src="documentation/images/snarePip.dag.png" width="700"/>
</p>

**DAG of Snare-seq automated processing system**

## Installation
Installation contains 2 parts, for meta-table and python related functions:
```bash
git clone https://github.com/huqiwen0313/snarePip.git
cd snarePip
pip install .
```

snarePip R package for atac-seq QC and downstream analysis
```bash
install.packages("devtools")
devtools::install_github("huqiwen0313/snarePip", ref="main")
```  

## Usage
For running the pipeline with default steps:
```bash
git clone https://github.com/huqiwen0313/snarePip.git
cd snarePip
python -m snarePip [arguments]
```
Arguments | Description
-----------|----------
`-s or --sampletable` | Name of google sheet contains sample information.
`-w or --worksheet` | which worksheet to load from google spreadsheet, default 0(the first worksheet).
`-r or --RNAdir` | path to RNA folder that processed results will be saved.
`-a or --ATACdir` | path to ATAC folder that processed results will be saved.
`-c or --cores` | number of cores used to run snakemake pipeline.
`-sr or --snakeRNA` | snakemake file for RNA processing.
`-sa or --snakeATAC` | snakemake file for ATAC processing.
`-t or --type` | assay type, e.g. snare_2, tenX.
